Makefile

```diff
@@ -142,6 +142,25 @@ frontend-lint: node_modules/.uptodate
 bddtests: python
 	@tox -qe bddtests
 
+PIP_COMPILE := pyenv exec pip-compile
+
+requirements/%.txt: requirements/%.in
+	$(PIP_COMPILE) --allow-unsafe --quiet $(PIP_COMPILE_FLAGS) $<
+
+requirements/dev.txt: requirements/requirements.txt
+requirements/tests.txt: requirements/requirements.txt
+requirements/functests.txt: requirements/requirements.txt
+requirements/bddtests.txt: requirements/requirements.txt
+requirements/lint.txt: requirements/tests.txt requirements/functests.txt requirements/bddtests.txt
+
+pip_compile_executable := $(shell pyenv root)/versions/$(shell pyenv version-name)/bin/pip-compile
+
+$(pip_compile_executable):
+	pyenv exec pip install --quiet --disable-pip-version-check pip-tools
```

Comment:
So this is installing an uncontrolled version of `pip-tools` at `~/.pyenv/versions/3.8.12/bin/pip-compile`. I think we might want to figure out how to get this pinned in a `.txt` file and updated by Dependabot. I think this may require `requirements/pipcompile.{in,txt}` files, a `.tox/pipcompile` venv, and `Makefile` running `pyenv exec tox -e pipcompile ...`. Note that `pip-compile` **doesn't need to be installed in the actual venvs** that the requirements files belong to, it just needs to be run with the right version of Python not necessarily in the same venv, it can and should be installed just once in its own separate venv.

Reply:
This seems to work for getting it to install and run `pip-compile` in each venv. It's actually simpler overall, the whole thing is just 11 lines of `Makefile` and installing `pip-tools` isn't a concern as it's just in the `*.in` files:

```diff
diff --git a/Makefile b/Makefile
index 32123521..3c1649dc 100644
--- a/Makefile
+++ b/Makefile
@@ -142,10 +142,8 @@ frontend-lint: node_modules/.uptodate
 bddtests: python
 	@tox -qe bddtests
 
-PIP_COMPILE := pyenv exec pip-compile
-
 requirements/%.txt: requirements/%.in
-	$(PIP_COMPILE) --allow-unsafe --quiet $(PIP_COMPILE_FLAGS) $<
+	@pyenv exec tox -qe $(subst requirements,dev,$(basename $(notdir $@))) --run-command 'pip-compile --allow-unsafe --quiet $(PIP_COMPILE_FLAGS) $<'
 
 requirements/dev.txt: requirements/requirements.txt
 requirements/tests.txt: requirements/requirements.txt
@@ -153,13 +151,8 @@ requirements/functests.txt: requirements/requirements.txt
 requirements/bddtests.txt: requirements/requirements.txt
 requirements/lint.txt: requirements/tests.txt requirements/functests.txt requirements/bddtests.txt
 
-pip_compile_executable := $(shell pyenv root)/versions/$(shell pyenv version-name)/bin/pip-compile
-
-$(pip_compile_executable):
-	pyenv exec pip install --quiet --disable-pip-version-check pip-tools
-
 .PHONY: requirements requirements/
-requirements requirements/: $(pip_compile_executable) $(wildcard requirements/*.txt)
+requirements requirements/: $(wildcard requirements/*.txt)
 
 .PHONY: sure
 sure: checkformatting backend-lint frontend-lint backend-tests coverage frontend-tests functests bddtests
diff --git a/requirements/bddtests.in b/requirements/bddtests.in
index 477dc2fa..8f2d6c0b 100644
--- a/requirements/bddtests.in
+++ b/requirements/bddtests.in
@@ -1,3 +1,4 @@
+pip-tools
 httpretty
 pytest
 factory-boy
diff --git a/requirements/coverage.in b/requirements/coverage.in
index 4ebc8aea..5c90533e 100644
--- a/requirements/coverage.in
+++ b/requirements/coverage.in
@@ -1 +1,2 @@
+pip-tools
 coverage
diff --git a/requirements/dev.in b/requirements/dev.in
index 7662cdb1..1b05378c 100644
--- a/requirements/dev.in
+++ b/requirements/dev.in
@@ -1,3 +1,4 @@
+pip-tools
 factory-boy
 pyramid-ipython
 supervisor
diff --git a/requirements/dockercompose.in b/requirements/dockercompose.in
index e41f7569..78495ea5 100644
--- a/requirements/dockercompose.in
+++ b/requirements/dockercompose.in
@@ -1 +1,2 @@
+pip-tools
 docker-compose
diff --git a/requirements/format.in b/requirements/format.in
index 7559a405..4fb9ec40 100644
--- a/requirements/format.in
+++ b/requirements/format.in
@@ -1,2 +1,3 @@
+pip-tools
 black
 isort
diff --git a/requirements/functests.in b/requirements/functests.in
index 9aca0452..c428af55 100644
--- a/requirements/functests.in
+++ b/requirements/functests.in
@@ -1,3 +1,4 @@
+pip-tools
 httpretty
 pytest
 factory-boy
diff --git a/requirements/lint.in b/requirements/lint.in
index 62660268..981db1c9 100644
--- a/requirements/lint.in
+++ b/requirements/lint.in
@@ -1,3 +1,4 @@
+pip-tools
 pylint
 pydocstyle
 pycodestyle
diff --git a/requirements/tests.in b/requirements/tests.in
index 49fcb127..fc938155 100644
--- a/requirements/tests.in
+++ b/requirements/tests.in
@@ -1,3 +1,4 @@
+pip-tools
 coverage
 httpretty
 aioresponses

```

For example:

```terminal
$ make --always-make -n requirements
pyenv exec tox -qe dev --run-command 'pip-compile --allow-unsafe --quiet  requirements/requirements.in'
pyenv exec tox -qe bddtests --run-command 'pip-compile --allow-unsafe --quiet  requirements/bddtests.in'
pyenv exec tox -qe coverage --run-command 'pip-compile --allow-unsafe --quiet  requirements/coverage.in'
pyenv exec tox -qe dev --run-command 'pip-compile --allow-unsafe --quiet  requirements/dev.in'
pyenv exec tox -qe dockercompose --run-command 'pip-compile --allow-unsafe --quiet  requirements/dockercompose.in'
pyenv exec tox -qe format --run-command 'pip-compile --allow-unsafe --quiet  requirements/format.in'
pyenv exec tox -qe functests --run-command 'pip-compile --allow-unsafe --quiet  requirements/functests.in'
pyenv exec tox -qe tests --run-command 'pip-compile --allow-unsafe --quiet  requirements/tests.in'
pyenv exec tox -qe lint --run-command 'pip-compile --allow-unsafe --quiet  requirements/lint.in'
```